go/ai/retriever.go

```diff
@@ -55,58 +71,154 @@ type RetrieverResponse struct {
-func LookupIndexer(provider, name string) *Indexer {
-	return (*Indexer)(core.LookupActionFor[*IndexerRequest, struct{}, struct{}](atype.Indexer, provider, name))
+func LookupIndexer(provider, name string) Indexer {
+	return (*IndexerAction)(core.LookupActionFor[*IndexerRequest, struct{}, struct{}](atype.Indexer, provider, name))
 }
 
 // DefineRetriever registers the given retrieve function as an action, and returns a
-// [Retriever] that runs it.
-func DefineRetriever(provider, name string, ret func(context.Context, *RetrieverRequest) (*RetrieverResponse, error)) *Retriever {
-	return (*Retriever)(core.DefineAction(provider, name, atype.Retriever, nil, ret))
+// [RetrieverAction] that runs it.
+func DefineRetriever(provider, name string, ret func(context.Context, *RetrieverRequest) (*RetrieverResponse, error)) *RetrieverAction {
+	return (*RetrieverAction)(core.DefineAction(provider, name, atype.Retriever, nil, ret))
 }
 
 // IsDefinedRetriever reports whether a [Retriever] is defined.
 func IsDefinedRetriever(provider, name string) bool {
-	return LookupRetriever(provider, name) != nil
+	return (*RetrieverAction)(core.LookupActionFor[*RetrieverRequest, *RetrieverResponse, struct{}](atype.Retriever, provider, name)) != nil
 }
 
 // LookupRetriever looks up a [Retriever] registered by [DefineRetriever].
 // It returns nil if the model was not defined.
-func LookupRetriever(provider, name string) *Retriever {
-	return (*Retriever)(core.LookupActionFor[*RetrieverRequest, *RetrieverResponse, struct{}](atype.Retriever, provider, name))
+func LookupRetriever(provider, name string) Retriever {
+	return (*RetrieverAction)(core.LookupActionFor[*RetrieverRequest, *RetrieverResponse, struct{}](atype.Retriever, provider, name))
 }
 
-// Index runs the given [Indexer].
-func (i *Indexer) Index(ctx context.Context, req *IndexerRequest) error {
+// Index runs the given [IndexerAction].
+func (i *IndexerAction) Index(ctx context.Context, req *IndexerRequest) error {
 	if i == nil {
 		return errors.New("Index called on a nil Indexer; check that all indexers are defined")
 	}
 	_, err := (*indexerAction)(i).Run(ctx, req, nil)
 	return err
 }
 
-// Retrieve runs the given [Retriever].
-func (r *Retriever) Retrieve(ctx context.Context, req *RetrieverRequest) (*RetrieverResponse, error) {
+// Retrieve runs the given [RetrieverAction].
+func (r *RetrieverAction) Retrieve(ctx context.Context, req *RetrieverRequest) (*RetrieverResponse, error) {
 	if r == nil {
 		return nil, errors.New("Retriever called on a nil Retriever; check that all retrievers are defined")
 	}
 	return (*retrieverAction)(r).Run(ctx, req, nil)
 }
 
-func (i *Indexer) Name() string { return (*indexerAction)(i).Name() }
+// retrieveParams represents various params of the Retrieve call.
+type retrieveParams struct {
+	request *RetrieverRequest
+}
+
+// generateOption configures params of the Retrieve call.
+type retrieveOption func(req *retrieveParams) error
```

Comment:
I think we should export this type, so there is a place to hang user-visible doc. I also think it will result in all the With... functions appearing together under RetrieveOption in the generated documentation.

Same for generate.go.

Reply:
I'm OK either way, but I believe you argued for these options (specifically the generate one) to be private before. 
Exported.